File: Shloka and the Broken Lift/d1.py
import sys
import math

input = sys.stdin.readline

T = int(input())

for _ in range(T):
    N = int(input())
    H = [0] + list(map(int, input().split()))
    
    reachable = [False] * (N + 1)
    reachable[1] = True
    
    for i in range(1, N + 1):
        if not reachable[i]:
            continue
        
        # Find all divisors of H[i] - these are possible jump lengths
        d = 1
        while d * d <= H[i]:
            if H[i] % d == 0:
                # d is a divisor
                if d >= 2:
                    j = i + d
                    if j <= N:
                        reachable[j] = True
                
                # H[i] // d is also a divisor
                other = H[i] // d
                if other >= 2 and other != d:
                    j = i + other
                    if j <= N:
                        reachable[j] = True
            
            d += 1
    
    print("YES" if reachable[N] else "NO")

File: Shloka and the Broken Lift/d1.java
import java.io.*;
import java.util.*;

public class d1 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine());
        
        while (T-- > 0) {
            int N = Integer.parseInt(br.readLine());
            int[] H = new int[N + 1];
            
            String[] input = br.readLine().split(" ");
            for (int i = 1; i <= N; i++) {
                H[i] = Integer.parseInt(input[i - 1]);
            }
            
            boolean[] reachable = new boolean[N + 1];
            reachable[1] = true;
            
            for (int i = 1; i <= N; i++) {
                if (!reachable[i]) continue;
                
                // Get divisors of H[i] - these are our possible jump lengths
                for (int d = 1; d * d <= H[i]; d++) {
                    if (H[i] % d == 0) {
                        if (d >= 2) {
                            int j = i + d;
                            if (j <= N) reachable[j] = true;
                        }
                        
                        int other = H[i] / d;
                        if (other >= 2 && other != d) {
                            int j = i + other;
                            if (j <= N) reachable[j] = true;
                        }
                    }
                }
            }
            
            System.out.println(reachable[N] ? "YES" : "NO");
        }
    }
}